pkg/services/ngalert/sender/router.go

```diff
@@ -128,8 +129,12 @@ func (d *AlertsRouter) SyncAndApplyConfigFromDatabase() error {
 		redactedAMs := buildRedactedAMs(d.logger, alertmanagers, cfg.OrgID)
 		d.logger.Debug("Alertmanagers found in the configuration", "alertmanagers", redactedAMs)
 
+		var urls []string
+		for _, cfg := range alertmanagers {
+			urls = append(urls, cfg.amURL)
+		}
 		// We have a running sender, check if we need to apply a new config.
-		amHash := asSHA256(alertmanagers)
+		amHash := asSHA256(append(urls, headersString(alertmanagers)))
```

Comment:
I found this very confusing initially because I couldn't understand why we were combining a header string with the URLs found. Then I realise that we have changed the signature of `alertmanagersFromDatasources` and everything made more sense.

However, `asSHA256` is only used here so it feels unnecessary complexity that we need to extract URLs and then have a separate function (in `headersString`) to combine both things that belong to a separate struct. 

I suggest you go the route of:

```golang
type externalAMcfg struct {
	amURL   string
	headers map[string]string
}

func (cfg *externalAMcfg) SHA256() string {
// loop through and re-use the implementation of `AsSHA256`
}
```

Reply:
I changed how this works, the hashes are generated based on the config and then just joined. This is way less confusing.